docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,485 @@
+
+Using sort operator with metric queries will result into an error.(e.g `count_over_time({job="prod/query-frontend"} | logfmt  | sort latency desc [5m])`)
+
+> __Warning:__ Sorting is dependent on how Grafana is currently ordering lines. Even if we re order our responses there is no guarantee that Grafana will be relying on it. And we might have to remove uncommon labels to keep only stream to ensure ordering across field streams. This means we will need a new API (`/v2`) to support ordering, although with compromise we might get something to work with the `v1` API.
+
+### Select
+
+The Select operator (`| select`) allows you to select and reduce field that should be returned.
+
+If you don't provide a select operator, all fields parsed will be returned by default. This means all fields from a log line if you're using implicit extraction. However if you used explicit extraction, only those fields will be returned.
+
+For example this query:
+
+```logql
+{job="prod/query-frontend"} | logfmt | sort duration desc | limit 25 | select query duration
+```
+
+is the same as this one
+
+```logql
+{job="prod/query-frontend"} | logfmt query duration | sort duration desc | limit 25
+```
+
+because the field extraction was explicit, only those fields (`query` and `duration`) are outputted.
+
+This is why select is important to use for large queries. Select will allows users to remove data (column) in their logs that are useless, repetitive and not interesting for the current investigation.
+
+The select operation will also support the [go templating language](https://golang.org/pkg/text/template/) when supplied with string argument and not fields.
+
+Fields extracted will be available under the `.FieldName` and `$FieldName` notation in the template.
+
+This way users can format each field to be more digestible. Returning a payload size  field with value of `12321421230123` is less meaningful than `12TB`.
+
+Example:
+
+```logql
+{job="prod/query-frontend"} | logfmt | select payload_size="{{ .size | humanize }}" level="{{ $level | ToUpper}}"
+```
+
+Renaming is allowed with or without templates, and is optional. An autogenerated field name will be used if it can't be inferred. (`field1`,`field2`, etc...)
+
+__In this current stance, select is not greedy meaning if you want to rename or template a single field, you'll have to declare all other fields to keep them.__
+
+In the future we might consider a `exclude` operator to take all fields but those passed as parameter. (`{job="prod/query-frontend"} | logfmt | exclude msg`)
+
+### Limit
+
+The limit operator (`| limit x`) is a new operator to overwrite the query string limit, this can be handy for building dashboard with different limits but also when writing and testing queries, you could start to experiment with a low limit to increase the feedback loop.
+
+If the limit operator is missing in the expression, the limit from the query string is used.
```

Comment:
I could see some confusion about the interaction of these two parameters.  Without looking at the code I wouldn't know when one or the other is applied.

Reply:
This is why the operator always takes precedence. I'll make sure this is well documented.